### PR TITLE
Add sorting functionality to flutter tables.

### PR DESCRIPTION
**Pull request by kenzieschmoll**
Screen Shot 2020-03-13 at 4 45 25 PM
Addresses part of https://github.com/flutter/devtools/issues/1657

### CODE REVIEW

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 1]
```diff
@@ -42,22 +45,28 @@ class FlatTable<T> extends StatefulWidget {
 
   final ItemCallback<T> onItemSelected;
 
+  final ColumnData<T> startingSortColumn;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart]: what does the word starting mean here? Can this just be sortColumn or can it just be a column index?

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 2]
```diff
@@ -102,6 +114,17 @@ class _FlatTableState<T> extends State<FlatTable<T>> {
...
+
+  void _sortDataAndUpdate(ColumnData column, SortDirection direction) {
+    setState(() {
+      sortData(column, direction);
+    });
+  }
+
+  @override
+  void sortData(ColumnData column, SortDirection direction) {
+    widget.data.sort((T a, T b) => _compareData(a, b, column, direction));
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart]: we shouldn't be modifying data on a stateless widget. that is a violation of the implicit assumptions for how a stateless widget works. instead have the stateful widget create its own sorted copy.

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 3]
```diff
@@ -146,25 +172,30 @@ class TreeTable<T extends TreeNode<T>> extends StatefulWidget {
...
   List<double> columnWidths;
   List<bool> rootsExpanded;
 
-  /// The number of items to show when animating out the tree table.
-  static const itemsToShowWhenAnimating = 50;
-
   @override
   void initState() {
     super.initState();
+    sortData(widget.startingSortColumn, widget.startingSortDirection);
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart]: cleanup the code that is duplicated in didUpdateWidget and initState. add a helper called in both places.

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 4]
```diff
@@ -173,6 +204,7 @@ class _TreeTableState<T extends TreeNode<T>> extends State<TreeTable<T>>
   @override
   void didUpdateWidget(TreeTable oldWidget) {
     super.didUpdateWidget(oldWidget);
+    sortData(widget.startingSortColumn, widget.startingSortDirection);
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart]: nit: this is too aggressive a didUpdateWidget. Only update when the widget actually changed.

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 5]
```diff
@@ -619,12 +704,29 @@ class _TableRowState<T> extends State<TableRow<T>>
...
+        content = InkWell(
+          onTap: () => _handleSortChange(column),
+          child: Row(
+            children: [
+              if (isSortColumn)
+                Icon(
+                  widget.sortDirection == SortDirection.ascending
+                      ? Icons.expand_less
+                      : Icons.expand_more,
+                  size: 16.0,
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart]: we shouldn't be leaking our default icon size all over the place. add this as a constant in the theme and cleanup other cases where we are hard coding size 16 assuming we are.

File: packages/devtools_app/lib/src/flutter/table.dart [hunk 6]
```diff
@@ -689,4 +791,27 @@ class _TableRowState<T> extends State<TableRow<T>>
...
+    } else if (columnData.numeric) {
+      direction = SortDirection.descending;
+    } else {
+      direction = SortDirection.ascending;
+    }
+    widget.onSortChanged(columnData, direction);
+  }
+}
+
+mixin TableSortMixin<T> {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/table.dart]: this is a bit of an odd mixin as the only mixed in functionality is private.
Consider making sortData part of an interface and moving the other two helpers as just private top level methods.

File: packages/devtools_app/lib/src/profiler/cpu_profile_controller.dart [hunk 7]
```diff
@@ -19,32 +19,30 @@ class CpuProfilerController {
   static CpuProfileData baseStateCpuProfileData = CpuProfileData.empty();
 
   /// Notifies that new cpu profile data is available.
-  ValueListenable get dataNotifier => _dataNotifier;
+  ValueListenable get data => _dataNotifier;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/profiler/cpu_profile_controller.dart]: this refactor looks only half done. The privates still have _notifier in the name. 
Also looks unrelated. Maybe move to a separate CL?

File: packages/devtools_app/lib/src/table_data.dart [hunk 8]
```diff
@@ -201,9 +201,9 @@ class TableData<T> extends Object {
     }
 
     if (_sortColumn == column) {
-      _sortDirection = _sortDirection == SortOrder.ascending
-          ? SortOrder.descending
-          : SortOrder.ascending;
+      _sortDirection = _sortDirection == SortDirection.ascending
```
Comment (jacob314) [inline packages/devtools_app/lib/src/table_data.dart]: add an extension method that reverses the sort direction.

Comment (jacob314) [inline packages/devtools_app/lib/src/table_data.dart]: use your extension method to reverse the sort direction.

File: packages/devtools_app/lib/src/utils.dart [hunk 9]
```diff
@@ -562,3 +562,16 @@ class Range {
...
   int get hashCode => hashValues(begin, end);
 }
+
+enum SortDirection {
+  ascending,
+  descending,
+}
+
+extension SortDirectionExtension on SortDirection {
+  SortDirection opposite() {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/utils.dart]: nit: would reverse() be clearer than opposite()?

File: packages/devtools_app/lib/src/flutter/table.dart (hunk 6 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/flutter/table.dart]: done

File: packages/devtools_app/lib/src/flutter/theme.dart [hunk 10]
```diff
@@ -39,6 +39,8 @@ ThemeData _lightTheme() {
 }
 
 const buttonMinWidth = 36.0;
+const defaultIconSize = 16.0;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/theme.dart]: fyi @devoncarew these constants might be relevant for some of your CLs as well.
